인증/인가부분

amm0124 opened the pull request:
## 해결하려는 문제가 무엇인가요? 
- Issue 12/13 

## 어떻게 해결했나요?

아직 하는 중입니다

## 더 하고 싶은 말

~일단 구현을 하고 있습니다. 로컬에 좀 있는데 아직 완성이 덜 되어가지고 일단 여기까지 PR 올립니다~

## 논의사항(거의 회고+다음에 해야할 것 짬뽕 )
(이라고 하고 그냥 떠들고 싶어서 씁니다!!)
그냥 여기에 주절주절 쓰겠습니다 ( 술 취한 것 아님)

1. accessToken 및 refreshToken 이녀석들은 프론트님들과 이야기를 해야겠어요. 쿠키 형식으로 주는지 등등
2. 만약 accessToken이 만료가 되었다면, 새로운 토큰을 발급해주는 것은 security단에서 해야 할까요? 아니면 service에서 해야 할까요? 진짜 어케 해야할지 잘 모르겠어요
3. 로그아웃 필터에는 redis와 같은 인메모리 db 사용해서 access Token과 refresh Token을 넣어줄 예정입니다.
4. 사실 더 빡세게 구현하라면 더 할 수 있긴 한데 그건 리팩토링 시간에 하도록 하겠습니다 . .. . . .
5. 그리고 공식 문서 보니까  Role << 이 녀석들 다 collection에 넣어서 관리하더라고요.. 다음에 만지작 
6. JwtAuthenticationProvider<<이라는 친구도 있는데 쉽게 말해서 인증객체 관리자더라고요 이 녀석이 Jwt Token을 decoding하는 역할을 하는 객체다 어쩌고 하던데 이 친구도 가능하면 써봐야겠습니다(AbstractAuthenticationProvider의 구현체)
 7. 아맞다 db에 password Encoding안하고 넣었어요 ㅇㄴ.  나중에 리팩토링 하겠습니다!! 일단 머지를 합시다 .. . .
 8. 로그인 하지 않고도 로직에 접근할 수 있는 Anonymous filter도 있더라고요. 참 많다 많아
 9. AbstractAuthenticationProcessingFilter의 구현체중 하나인 OAuth2LoginAuthenticationFilter도 있더라고요. 이를 통해서 카카오 로그인을 하는 것은!

Commit:
build : spring security 의존성 추가

Commit:
chore : application-secret.yml 작성

Commit:
feat : security config 기본 작성
TODO
- Oauth2.0 인증인가
- error handling을 위한 AuthenticationEntryPoint , AccessDeniedException 구현
- AbstractUserDetailsAuthenticationProvider의 구현체 DaoAuthenticationFilter 만들어야 함
- 회원 가입

Commit:
build : 모니터링을 위한 actuator 의존성 추가

Commit:
feat : member error 추가

Commit:
feat : member entity 추가
일단 대충 구현했습니다. member entity에 어떤 column이 추가 될 지 아직 확정이 안 나서 기본값으로 넣었습니다!
그리고 로그인 되는 거 명시적으로 보려고 타임리프 써서 login page 임시로 넣었습니다~

Commit:
build : jjwt 버전 변경

stopmin commented on src/main/resources/application-secret.yml:
```diff
@@ -0,0 +1,9 @@
+kakao :
```
이거 secret PR에서 제외시켜야할 것 같습니다.

amm0124 replied:
.gitignore을 조금 나중에 작성해가지고 저게 올라왔네요 ㅠ.ㅠ 
나중에 key 발급 받은 후엔 빼겠습니당

stopmin replied:
넵~~ 좋습니다! 지금은 별거없어서 상관없을 것 같아요!! 
혹시 모르는 것 같아서 말씀드렸습니다!~~

canyos commented:
코드가 맛있네요

Commit:
feat : 로그인 성공 시, "/" redirect하는 handler 구현
- 경로는 차후 바뀜 (임시 경로)

Commit:
feat : 로그인 실패 시, "/" redirect하는 handler 구현
- 경로는 차후 바뀜 (임시 경로)

Commit:
refactoring : success handler log 추가

Commit:
feat : 로그아웃 성공시, "/" redirect하는 handler 구현
- 경로는 차후 바뀜 (임시 경로)

stopmin commented on src/main/java/poomasi/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,38 @@
+    @Column(nullable = true)
+    private Role role;
+
+    @Enumerated(EnumType.STRING)
```
오오 좋습니다~~

stopmin commented on src/main/java/poomasi/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,38 @@
+    private LoginType loginType;
+
+    @Column(nullable = true)
+    private String snsAuthId;
```
혹시 snsAuthId는 따로 인덱싱 설정이 필요하지 않은가 궁금합니다~~!! (진짜 궁금)

amm0124 replied:
일단아직까지는 생각안했습니다람쥐!! 임시로써둔것

stopmin commented on src/main/java/poomasi/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,38 @@
+    private Long id;
+
+    @Column(unique = true, nullable = true, length = 50)
+    private String email;
```
알고계실 것 같긴 합니다만 나중에 컨트롤러 단에서 이메일 형식이 올바른지 validation하는 로직을 넣어도 좋을 것 같아욤~~!! (혹시나 클라이언트에게 요청을 받는거라면? 해당 사항 관련해서  아래 코드 참고하시면 될거같습니다!!!)
```
@NotBlank(message = "이메일은 필수 항목입니다.")
@Email(message = "유효한 이메일 형식이어야 합니다.")
private String email;
```

stopmin commented on src/main/java/poomasi/domain/member/entity/Role.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.security.core.GrantedAuthority;
+
+
+//extends GrantedAuthority를 해서 구현한 것도 있던데 . . . 좀 더 찾아봐야겠어요
```
👍

Commit:
chore : yml 작성

Commit:
refactoring : Member column email -> username 변경

Commit:
feat : member repository 메서드 추가

Commit:
feat : Role을 String 형태로 return하는 메서드 추가

Commit:
refactoring : Security Filter chain 변경
- csrf 해제
- cors 해제
- jwt login을 위한 session stateless
- jwt 인증 필터 UsernamePasswordAuthenticationFilter앞 추가
- 로그인 필터 추가
- 기본 폼 로그인, http basic 해제
- 로그아웃 필터

Commit:
feat : 임시 필터 구현. PR후 완성 예정

Commit:
feat : AccessDeniedHandler 구현

Commit:
feat : 인증 객체(Authentication Provider)를 위한
UserDetail, UserDetailService 구현

Commit:
feat : 스프링 빈 등록
- password encoder(for Authentication Provider)
- pattern builder(whitelist)

Commit:
chore : 기타 사항 push

Commit:
Merge remote-tracking branch 'origin/week4' into feature/ISSUE-12

# Conflicts:
#	.gitignore
#	src/main/java/poomasi/global/error/BusinessError.java

Commit:
chore : conflict 해결

jjt4515 commented on src/main/java/poomasi/domain/auth/customFilter/LogoutFilterImpl.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        //blacklist에 access token, refresh token 추가 (만료기간까지 저장해야 함)
```
blacklist에는 access token만 추가하면 됩니다. 보안 상의 이유로 refresh token은 로그인 할 때 redis에 넣고, 로그아웃 시 삭제를 해야 하기 때문에 blacklist에는 넣을 필요가 없습니다. 로그인 때 넣는 이유는 다음과 같습니다.
1.유저가 리프레시 토큰 탈취 당함 -> 유저가 토큰 갱신 -> 탈취자가 리프레시 토큰 사용 -> 이전 리프레시 토큰(레디스에 저장된 거)이랑 다른 거 확인하고 로그아웃
2.유저가 리프레시 토큰 탈취 당함 -> 탈취자가 토큰 갱신 -> 유저가 리프레시 토큰 사용 -> 이전 리프레시 토큰(레디스에 저장된 거)이랑 다른 거 확인하고 로그아웃

stopmin replied:
제 PR은 아니지만 개인적으로 궁금한 부분이 있어 남깁니다 :D
제가 잘 이해했는지 궁금하고, 같이 이야기 해보면 좋을 것 같아서 남겨욤!!!

### 1️⃣ 보안상의 이유로 Refresh Token을 로그인할 때 Redis에 저장하고, 로그아웃 시 삭제한다는 점에 대해 궁금한 점이 있습니다. 
블랙리스트에서 Access Token만 필요하다는 설명이 이해되지 않습니다. 만약 Refresh Token이 탈취된다면, 그 토큰이 더 이상 유효하지 않도록 반드시 블랙리스트에 넣어 무효화해야 하지 않나요? 블랙리스트의 주된 목적은 탈취된 토큰이 재사용되지 않도록 막기 위한 수단인데, Refresh Token이 탈취된다면 이를 무효화하지 않는 것이 과연 의미가 있을까요?

### 2️⃣ 로그아웃 시 레디스에 저장한 토큰을 삭제한다는 논리가 나온 동기가 궁금합니다!
로그아웃시 Redis에 토큰을 저장하지 않는다면, 로그아웃 후에 탈취된 토큰이 Redis에 없기 때문에 재로그인을 요구하겠다는 것이 목적이라면, 이는 논리적으로 몇 가지 모순이 있다고 생각합니다. (물론 제가 틀렸을 수도 있습니다🥺)

첫째, 논리적으로 모순된 이유는, **정상적인 로그인 시도 시에도 매번 블랙리스트에 토큰을 넣게 되면 새로 발급된 유효한 토큰이 블랙리스트에 들어가 사용할 수 없게 됩니다**. 매번 로그인이 성공하더라도 블랙리스트에 들어가기 때문에, 토큰이 새로 발급되었음에도 불구하고 무효화되는 상황이 발생하게 되는 것이죠.

둘째, **로그인의 의미가 상실**됩니다. 로그인은 사용자가 인증을 통해 새로운 유효한 토큰을 발급받는 과정입니다. 그런데 로그인할 때마다 그 토큰을 블랙리스트에 넣거나 무효화한다면, 발급된 토큰을 사용할 수 없게 되어 로그인 자체의 의미가 퇴색됩니다.

본래 블랙리스트는 유효 기간이 남아있는 토큰 중에서 악의적으로 사용될 가능성이 있는 토큰을 차단하기 위한 수단입니다. 따라서 일반적으로는 로그아웃 시에 Access Token과 Refresh Token 모두를 블랙리스트에 추가하여 더 이상 유효하지 않도록 차단하는 것이 적절합니다.

### 3️⃣ 혹시나 화이트리스트 방식을 말씀하고 싶으신건지 궁금합니다!!
혹시나 화이트리스트 방식으로 로그인을 할 때 Redis에 토큰을 저장하고, 로그아웃할 때 삭제하는 화이트리스트 방식으로 관리한다면, 먼저 Redis에 대한 의존성이 매우 커집니다.  

Redis에 장애가 발생할 경우 인증 시스템 전체에 영향을 미칠 수 있고, Redis에 저장되는 데이터도 많아지면서 메모리 사용량이 증가해 추가적인 캐시 관리가 필요해질 수 있습니다. 토큰 갱신 처리 역시 Access Token과 Refresh Token을 함께 다루어야 하므로 더 복잡해지며, 화이트리스트 방식에서는 토큰 만료 관리도 어려워질 수 있습니다. 진택님이 언급하신 방식은 이러한 화이트리스트 방식과 유사한 것 같습니다.


## 이후 결론...
> 너무 궁금해서 건호님께 물어본 결과..!

1. access Token은 기존의 방식과 동일하게 블랙리스트 방식대로 로그아웃하면 저장하고, 일정 시간 후에 날리는게 맞았다!
2. refresh Token은 자동 로그인을 위해서(?) 남겨준다.

이후에 건호님이 문서화 해주신다고 하셔서 노션에 같이 보면 좋을 것 같아요!! (어렵네유..)

jjt4515 replied:
1. 리프레시 토큰은 레디스에 직접 저장하기 때문에 블랙 리스트에 넣어서 검사할 필요가 없어집니다. 로그아웃 시 레디스에서 삭제되기 때문에 이 자체가 무효화 과정입니다.
2. 블랙 리스트에는 액세스 토큰 만을 넣는데, 이는 로그아웃 시에만 넣습니다. 따라서 로그인 시 생성되는 액세스 토큰이 무효화 되진 않습니다.
3. 오 의존성이 커지는 문제는 있겠네요

액세스 토큰과 리프레시 토큰을 블랙 리스트로 관리할지 화이트 리스트(?)로 관리할지는 선택으로 보입니다. 저는 아래 이유로 액세스 토큰은 블랙 리스트로 관리, 리프레시 토큰은 레디스에 저장해서 관리하는 방식을 선택하였습니다.

만약 탈취자와 사용자가 모두 로그인 상태일 때, 탈취 당하면 블랙 리스트 방식으로는 해결할 수 없습니다. 

그래서 저는 유효 기간이 긴 리프레시 토큰에 대해서는 로그인 되어 있는 동안에 리프레시 토큰이 레디스에 저장되는 방식을 사용하였습니다. 이는 탈취자가 리프레시 토큰을 갱신해서 사용할 때, 사용자도 갱신을 시도한다면 레디스에 저장되어있는 갱신 상태의 리프레시 토큰과 값이 다르기에 로그아웃 처리를 함으로써 탈취를 해결하는 효과가 있습니다. 또 건호님 말대로 자동 로그인 시 유효하다는 장점이 있습니다. 

유효 기간이 짧은 액세스 토큰에 대해서는 5분이라는 유효기간은 로그인 중에 탈취 당해도 부담이 적다는 점을 고려하여 
 지민 님이 말씀하신 의존성, 비용 등의 문제로 블랙 리스트 방식을 사용하였습니다.
